File: leetcode/spiralMatrix2.py
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        
        matrix = [[0 for i in range(n)] for j in range(n)]
        
        directions = [[0, 1], [1, 0], [0, -1], [-1, 0]]
        
        s = 0
        i = 0
        j = 0
        
        for k in range(1, n*n+1):
            matrix[i][j] = k
            i += directions[s][0]
            j += directions[s][1]
            if(i < 0 or i >= n or j < 0 or j >= n or matrix[i][j] != 0):
                i -= directions[s][0]
                j -= directions[s][1]
                s = (s+1)%4
                i += directions[s][0]
                j += directions[s][1]
                
        return matrix
            
            

File: leetcode/spiralMatrix2.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> ans(n,vector<int>(n,0));
        
        vector<int> dir{0, 1, 0, -1};
        
        int s = 0;
        int i = 0;
        int j = 0;
        
        for(int k=1; k<=n*n; k++) {
            ans[i][j] = k;
            i += dir[s];
            j += dir[(s+1)%4];
            if(i<0 || i==n || j<0 || j==n || ans[i][j]!=0) {
                i -= dir[s];
                j -= dir[(s+1)%4];
                s = (s+1)%4;
                i += dir[s];
                j += dir[(s+1)%4];
            }
        }
        
        return ans;
    }
};